pkg/actions/db_service.go

```diff
@@ -402,3 +443,185 @@ func (s *DBService) NewRunID() string {
+			data, err := proto.Marshal(&sec)
+			if err != nil {
+				return err
+			}
+
+			if err = encoder.Encode(kv.Entry{
+				PartitionKey: []byte(PartitionKey),
+				Key:          secKey,
+				Value:        data,
+			}); err != nil {
+				return err
+			}
+		}
+
+		// Add secondary keys commit
+		if m.CommitID != "" {
+			secKey := []byte(RunByCommitPath(m.RepoID, m.CommitID, m.RunResult.RunID))
+			sec := kv.SecondaryIndex{PrimaryKey: key}
+			data, err := proto.Marshal(&sec)
+			if err != nil {
+				return err
+			}
+			if err = encoder.Encode(kv.Entry{
+				PartitionKey: []byte(PartitionKey),
+				Key:          secKey,
+				Value:        data,
+			}); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// tasksWorker dispatch taskWorker jobs per run
+func tasksWorker(ctx context.Context, d *pgxpool.Pool, jobChan <-chan *migrateTask) error {
+	tasksChan := make(chan *taskResultMigrate, migrateQueueSize)
+	var g multierror.Group
+	for i := 0; i < taskWorkers; i++ {
+		g.Go(func() error {
+			return taskWorker(tasksChan)
+		})
+	}
+	for t := range jobChan {
+		tasksRows, err := d.Query(ctx, "SELECT * FROM actions_run_hooks where run_id=$1", t.oldID)
+		if err != nil {
+			return err
+		}
+		taskScanner := pgxscan.NewRowScanner(tasksRows)
+		for tasksRows.Next() {
+			tm := new(taskResultMigrate)
+			err = taskScanner.Scan(tm)
+			if err != nil {
+				return err
+			}
+			tm.RunID = t.newID
+			tasksChan <- tm
+		}
+		tasksRows.Close()
+	}
+	close(tasksChan)
+
+	return g.Wait().ErrorOrNil()
+}
+
+// taskWorker writing task data to fd
+func taskWorker(jobChan <-chan *taskResultMigrate) error {
+	for t := range jobChan {
+		taskPr := protoFromTaskResult(&t.TaskResult)
+		taskValue, err := proto.Marshal(taskPr)
+		if err != nil {
+			return err
+		}
+		taskKey := []byte(kv.FormatPath(TasksPath(t.RepoID, t.RunID), t.HookRunID))
+		if err = encoder.Encode(kv.Entry{
+			PartitionKey: []byte(PartitionKey),
+			Key:          taskKey,
+			Value:        taskValue,
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func Migrate(ctx context.Context, d *pgxpool.Pool, writer io.Writer) error {
+	encoder = safeEncoder{
+		je: json.NewEncoder(writer),
+		mu: sync.Mutex{},
+	}
+	// Create header
+	if err := encoder.Encode(kv.Header{
+		LakeFSVersion:   version.Version,
+		PackageName:     packageName,
+		DBSchemaVersion: kv.InitialMigrateVersion,
+		CreatedAt:       time.Now().UTC(),
+	}); err != nil {
+		return err
+	}
+
+	runChan := make(chan *runResultMigrate, migrateQueueSize)
+	tasksChan := make(chan *migrateTask, migrateQueueSize)
+	var g multierror.Group
+	for i := 0; i < runWorkers; i++ {
+		// Run worker reads runs from DB and writes respected entries to fd
+		g.Go(func() error {
+			return runWorker(runChan)
+		})
+		// tasks worker scans tasks per run ID and starts the task go routines which write the respected entries to fd
+		g.Go(func() error {
+			return tasksWorker(ctx, d, tasksChan)
+		})
```

Comment:
I think it could be simplified if we had a single type that would handle a run and all its hooks. Right now it's hard to choose the right amount of workers of each type. You choose to create equal amount, but is the work really distributed evenly?

Reply:
There's a question of the ratio of total runs vs total tasks, which I'm not sure there a singular truth to. I played with it a little with the worker counts and my conclusion was that the impact is not substantial with the given dataset (ratio of 10 tasks per run)